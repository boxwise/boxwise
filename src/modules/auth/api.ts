/* eslint-disable @typescript-eslint/no-non-null-assertion */

import { firebase, db } from "firebaseFactory";
import { Flavor } from "typeScriptExtensions";

export type GetCurrentUser = () => Promise<User>;

export type OrganizationRef = Flavor<string, "Organization">;
export type UserProfileRef = Flavor<string, "UserProfile">;

export interface User {
  uid: string;
  userProfileRef: UserProfileRef;
  emailAddress: string;
  displayName: string;
  organizationName: string;
  organizationRef: OrganizationRef;
}

export const signOut = (): Promise<void> => {
  return firebase.auth().signOut();
};

const getUserFromData = (firebaseUser: firebase.User): Promise<User> => {
  if (!firebaseUser.email) throw new Error("User doesn't have email specified");
  // ideally we'd have all the information we need sitting on the profile
  // but for now we hide the complexity (and slow-ness) here
  return db
    .collection("profiles")
    .doc(firebaseUser.uid)
    .get()
    .then(profileDoc => {
      const profileData = profileDoc.data();
      if (!profileData) throw new Error("User doesn't have profile available");
      return profileData.organization
        .get()
        .then((orgDoc: firebase.firestore.DocumentSnapshot) => ({
          uid: firebaseUser.uid,
          emailAddress: firebaseUser.email,
          displayName: profileData.name,
          userProfileRef: `profiles/${profileDoc.id}`,
          organizationName: orgDoc.data()!.name,
          organizationRef: `organizations/${orgDoc.id}`
        }));
    });
};

export const signIn = (email: string, password: string): Promise<void> => {
  return firebase
    .auth()
    .signInWithEmailAndPassword(email, password)
    .then(credential => Promise.resolve());
};

export const sendPasswordResetEmail = (email: string): Promise<void> => {
  return firebase.auth().sendPasswordResetEmail(email);
};

export const changePassword = async (
  getCurrentUser: GetCurrentUser,
  currentPassword: string,
  newPassword: string
): Promise<void> => {
  const { emailAddress } = await getCurrentUser();
  return firebase
    .auth()
    .signInWithEmailAndPassword(emailAddress, currentPassword)
    .then(({ user }) => user!.updatePassword(newPassword));
};

export const observeAuthState = (
  onSignOut: () => void,
  onSignIn: (getUserProfile: () => Promise<User>) => void
) => {
  return firebase.auth().onAuthStateChanged(user => {
    if (!user) {
      onSignOut();
    } else {
      onSignIn(() => getUserFromData(user));
    }
  });
};
