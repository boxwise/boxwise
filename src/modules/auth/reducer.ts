import { RootAction } from "storeTypes";

import {
  PASSWORD_EDIT,
  PASSWORD_RESET,
  USER_NOT_SIGN_IN,
  USER_SIGN_IN,
  USER_SIGN_OUT
} from "./actions";

export default function user(
  state = {
    loading: null,
    hasTriggeredReset: false,
    data: null,
    error: null
  },
  { type, payload }: RootAction
) {
  switch (type) {
    case USER_SIGN_IN.START:
      return { ...state, error: null, loading: true };

    case USER_SIGN_IN.SUCCESS:
      return { ...state, loading: false, data: payload };

    case USER_SIGN_IN.ERROR:
      return { ...state, loading: false, error: payload };

    case USER_NOT_SIGN_IN:
      return { ...state, loading: false };

    case USER_SIGN_OUT:
      return { ...state, data: null };

    case PASSWORD_RESET.START:
      return { ...state, loading: true, hasTriggeredReset: false };

    case PASSWORD_RESET.SUCCESS:
      return { ...state, loading: false, hasTriggeredReset: true };

    case PASSWORD_RESET.ERROR:
      return {
        ...state,
        loading: false,
        hasTriggeredReset: false,
        error: payload
      };

    case PASSWORD_EDIT.START:
      return { ...state, isUpdating: true, error: null };

    case PASSWORD_EDIT.SUCCESS:
      return { ...state, isUpdating: false, error: null };

    case PASSWORD_EDIT.ERROR:
      return { ...state, isUpdating: false, error: payload };

    default:
      return state;
  }
}
