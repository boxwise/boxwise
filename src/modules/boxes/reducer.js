import { BOX_ADD_ERROR, BOX_ADD_START, BOX_ADD_SUCCESS } from "./actions";

export default function boxes(
  state = { data: [], loading: false, error: null },
  { type, payload }
) {
  switch (type) {
    case BOX_ADD_START:
      return { ...state, loading: true };

    case BOX_ADD_SUCCESS:
      return { ...state, loading: false, data: [...state.data, payload] };

    case BOX_ADD_ERROR:
      return { ...state, loading: false, error: payload };

    default:
      return state;
  }
}
