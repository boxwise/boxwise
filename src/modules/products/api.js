import { firebase, db } from "firebaseFactory";

const getProductFromData = doc => {
  const data = doc.data();
  // we don't want to just store raw firebase data in here
  // else you get a load of firebase variables in the redux store
  return {
    id: doc.id,
    name: data.name,
    category: data.category,
    organizationId: data.organization.id,
    createdAt: data.createdAt,
    createdById: data.createdBy.id,
    isDeleted: data.isDeleted
  };
};

export const getAllProducts = async getCurrentUser => {
  const { organizationRef } = await getCurrentUser();

  return db
    .collection("products")
    .where("organization", "==", db.doc(organizationRef))
    .where("isDeleted", "==", false)
    .orderBy("category", "asc")
    .orderBy("name", "asc")
    .get()
    .then(({ docs }) => {
      return docs.map(getProductFromData);
    });
};

export const addProduct = async (product, getCurrentUser) => {
  const { organizationRef, userProfileRef } = await getCurrentUser();

  const values = {
    ...product,
    organization: db.doc(organizationRef),
    createdAt: firebase.firestore.FieldValue.serverTimestamp(),
    createdBy: db.doc(userProfileRef),
    isDeleted: false
  };

  return db
    .collection("products")
    .add(values)
    .then(ref => ref.get())
    .then(getProductFromData);
};

export const updateProduct = product => {
  const ref = db.collection("products").doc(product.id);

  return ref
    .update(product)
    .then(() => ref.get())
    .then(getProductFromData);
};

export const deleteProduct = productId => {
  return db
    .collection("products")
    .doc(productId)
    .update({ isDeleted: true });
};
