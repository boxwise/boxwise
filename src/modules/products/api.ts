import { firebase, db } from "firebaseFactory";
import { Flavor } from "typeScriptExtensions";
import {
  GetCurrentUser,
  OrganizationRef,
  UserProfileRef
} from "modules/auth/api";

export type ProductId = Flavor<string, "Product">;

export interface Product {
  id: ProductId;
  name: string;
  category: string;
  organizationId: OrganizationRef;
  createdAt: string;
  createdById: UserProfileRef;
  isDeleted: boolean;
}

const getProductFromData = (
  doc: firebase.firestore.DocumentSnapshot
): Product => {
  const data = doc.data();
  if (!data) throw new Error(`No data available on ${doc.ref}`);
  // we don't want to just store raw firebase data in here
  // else you get a load of firebase variables in the redux store
  return {
    id: doc.id,
    name: data.name,
    category: data.category,
    organizationId: data.organization.id,
    createdAt: data.createdAt,
    createdById: data.createdBy.id,
    isDeleted: data.isDeleted
  };
};

export const getAllProducts = async (
  getCurrentUser: GetCurrentUser
): Promise<Product[]> => {
  const { organizationRef } = await getCurrentUser();
  return db
    .collection("products")
    .where("organization", "==", db.doc(organizationRef))
    .where("isDeleted", "==", false)
    .orderBy("category", "asc")
    .orderBy("name", "asc")
    .get()
    .then(({ docs }) => {
      return docs.map(getProductFromData);
    });
};

export const addProduct = async (
  product: {
    name: string;
    category: string;
  },
  getCurrentUser: GetCurrentUser
): Promise<Product> => {
  const { organizationRef, userProfileRef } = await getCurrentUser();

  const values = {
    ...product,
    organization: db.doc(organizationRef),
    createdAt: firebase.firestore.FieldValue.serverTimestamp(),
    createdBy: db.doc(userProfileRef),
    isDeleted: false
  };

  return db
    .collection("products")
    .add(values)
    .then(ref => ref.get())
    .then(getProductFromData);
};

export const updateProduct = (product: Product): Promise<Product> => {
  const ref = db.collection("products").doc(product.id);

  return ref
    .update(product)
    .then(() => ref.get())
    .then(getProductFromData);
};

export const deleteProduct = (productId: ProductId): Promise<void> => {
  return db
    .collection("products")
    .doc(productId)
    .update({ isDeleted: true });
};
