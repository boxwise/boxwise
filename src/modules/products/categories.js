export const CATEGORY_MAN = "Man";
export const CATEGORY_WOMAN = "Woman";
export const CATEGORY_ADULT = "Adult";
export const CATEGORY_BOY = "Boy";
export const CATEGORY_GIRL = "Girl";
export const CATEGORY_CHILD = "Child";
export const CATEGORY_BABY = "Baby";
export const CATEGORY_FOOD = "Food";
export const CATEGORY_HYGIENE = "Hygiene";
export const CATEGORY_OTHER = "Other";

export const CATEGORIES = [
  CATEGORY_MAN,
  CATEGORY_WOMAN,
  CATEGORY_ADULT,
  CATEGORY_BOY,
  CATEGORY_GIRL,
  CATEGORY_CHILD,
  CATEGORY_BABY,
  CATEGORY_FOOD,
  CATEGORY_HYGIENE,
  CATEGORY_OTHER
];
