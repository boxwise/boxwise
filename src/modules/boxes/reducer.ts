import { RootAction, RootState, BoxesState } from "redux/storeTypes";
import { createReducerForIndexedState } from "redux/reducerFactory";

import { BOX_ADD, BOX_LIST } from "./actions";
import { Box } from "./api";

export interface BoxWithProductInfo extends Box {
  productName: string;
  productCategory: string;
}

export const getBoxesWithProductInfoFromState = ({
  boxes,
  products
}: RootState): {
  loading: boolean;
  error?: string;
  data: BoxWithProductInfo[];
} => {
  const isLoading = products.loading || boxes.loading;
  const data = {
    loading: isLoading,
    error: boxes.error,
    data: isLoading
      ? []
      : boxes.allIds.map(boxId => {
          const box = boxes.byId[boxId];
          const product = products.byId[box.productId];
          return {
            ...box,
            productName: (product && product.name) || "!!!Missing product",
            productCategory:
              (product && product.category) || "!!!Missing product"
          };
        })
  };
  return data;
};

const boxesIndexedStateReducer = createReducerForIndexedState<Box, BoxesState>(
  BOX_LIST,
  BOX_ADD
);

export default function boxes(
  state: BoxesState = {
    byId: {},
    allIds: [],
    loading: false,
    error: undefined
  },
  action: RootAction
) {
  const { type } = action;

  switch (type) {
    default:
      return boxesIndexedStateReducer(state, action);
  }
}
