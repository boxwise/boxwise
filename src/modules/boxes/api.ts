import { firebase, db } from "firebaseFactory";
import { GetCurrentUser } from "modules/auth/api";
import { ProductId } from "modules/products/api";
import { Flavor } from "typeScriptExtensions";

export type BoxId = Flavor<string, "Box">;

export interface Box {
  id: BoxId;
  quantity: number;
  comment?: string;
  productId: string;
  organizationId: string;
  createdAt: string;
  createdById: string;
  humanId: string;
}

const getBoxFromData = (doc: firebase.firestore.DocumentSnapshot): Box => {
  const data = doc.data();
  if (!data) throw new Error(`Could not load data from ${doc.ref}`);
  // we don't want to just store raw firebase data in here
  // else you get a load of firebase variables in the redux store
  return {
    id: doc.id,
    quantity: data.quantity,
    comment: data.comment,
    productId: data.product.id,
    organizationId: data.organization.id,
    createdAt: data.createdAt,
    createdById: data.createdBy.id,
    humanId: data.humanID
  };
};

// TODO: want to subscribe to updates
export const fetchActiveBoxes = async (getCurrentUser: GetCurrentUser) => {
  const { organizationRef } = await getCurrentUser();
  // TODO: we only want to return boxes that are currently 'active'
  return db
    .collection("boxes")
    .where("organization", "==", db.doc(organizationRef))
    .orderBy("humanID", "asc")
    .get()
    .then(({ docs }) => {
      return docs.map(getBoxFromData);
    });
};

export const addBox = async (
  quantity: number,
  comment: string,
  productId: ProductId,
  getCurrentUser: GetCurrentUser
) => {
  const { organizationRef, userProfileRef } = await getCurrentUser();
  const box = {
    quantity,
    comment: comment || "",
    organization: db.doc(organizationRef),
    product: db.doc(`products/${productId}`),
    createdAt: firebase.firestore.FieldValue.serverTimestamp(),
    createdBy: db.doc(userProfileRef),
    humanID: Math.floor(Math.random() * 1000000)
  };

  return db
    .collection("boxes")
    .add(box)
    .then(ref => ref.get())
    .then(getBoxFromData);
};
