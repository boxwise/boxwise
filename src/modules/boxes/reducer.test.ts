/* eslint-disable @typescript-eslint/no-non-null-assertion */
/* eslint-disable @typescript-eslint/no-object-literal-type-assertion */

import { RootState, BoxesState, ProductsState } from "redux/storeTypes";
import { CATEGORY_ADULT } from "modules/products/categories";
import { Product } from "modules/products/api";

import { Box } from "./api";
import { getBoxesWithProductInfoFromState } from "./reducer";

// firebase mock doesn't support querying by reference
describe("getBoxesWithProductInfoFromState", () => {
  const boxes: Partial<BoxesState> = {
    allIds: ["123", "456"],
    byId: {
      "123": {
        id: "123",
        productId: "AdultShorts",
        quantity: 100,
        humanId: "213"
      } as Box,
      "456": {
        id: "456",
        productId: "AdultTShirt",
        quantity: 50,
        humanId: "213"
      } as Box
    }
  };

  const products: Partial<ProductsState> = {
    byId: {
      AdultTShirt: {
        name: "Adult T-Shirts",
        category: CATEGORY_ADULT
      } as Product,
      AdultShorts: { name: "Adult Shorts", category: CATEGORY_ADULT } as Product
    }
  };

  const result = getBoxesWithProductInfoFromState({
    boxes,
    products
  } as RootState);

  it("should join boxes with products", () => {
    expect(result.data).toHaveLength(2);
    expect(result.data[0].productCategory).toEqual(
      products!.byId!.AdultTShirt.category
    );
  });

  it("should have consistent properties", () => {
    expect(result).toMatchSnapshot();
  });
});
