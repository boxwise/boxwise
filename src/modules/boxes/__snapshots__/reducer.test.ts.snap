// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getBoxesWithProductInfoFromState should have consistent properties 1`] = `
Object {
  "data": Array [
    Object {
      "humanId": "213",
      "id": "123",
      "productCategory": "Adult",
      "productId": "AdultShorts",
      "productName": "Adult Shorts",
      "quantity": 100,
    },
    Object {
      "humanId": "213",
      "id": "456",
      "productCategory": "Adult",
      "productId": "AdultTShirt",
      "productName": "Adult T-Shirts",
      "quantity": 50,
    },
  ],
  "error": undefined,
  "loading": undefined,
}
`;
